search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/prompt/PromptUtil.java

```diff
@@ -134,37 +146,132 @@ public static String buildSingleStringPrompt(
         return bldr.toString();
     }
 
+    /**
+     * Message APIs such as OpenAI's Chat Completion API and Anthropic's Messages API
+     * use an array of messages as input to the LLM and they are better suited for
+     * multi-modal interactions using text and images.
+     *
+     * @param provider
+     * @param systemPrompt
+     * @param userInstructions
+     * @param question
+     * @param chatHistory
+     * @param contexts
+     * @return
+     */
     @VisibleForTesting
     static String buildMessageParameter(
+        Llm.ModelProvider provider,
         String systemPrompt,
         String userInstructions,
         String question,
         List<Interaction> chatHistory,
         List<String> contexts
     ) {
+        return buildMessageParameter(provider, systemPrompt, userInstructions, question, chatHistory, contexts, null);
+    }
+
+    static String buildMessageParameter(
+        Llm.ModelProvider provider,
+        String systemPrompt,
+        String userInstructions,
+        String question,
+        List<Interaction> chatHistory,
+        List<String> contexts,
+        List<MessageBlock> llmMessages
+    ) {
 
         // TODO better prompt template management is needed here.
 
         if (Strings.isNullOrEmpty(systemPrompt) && Strings.isNullOrEmpty(userInstructions)) {
-            systemPrompt = DEFAULT_SYSTEM_PROMPT;
+            // Some model providers such as Anthropic do not allow the system prompt as part of the message body.
+            userInstructions = DEFAULT_SYSTEM_PROMPT;
         }
 
-        JsonArray messageArray = new JsonArray();
+        MessageArrayBuilder bldr = new MessageArrayBuilder(provider);
```

Comment:
`bldr` is hard to understand , use more meaningful variable name?

Reply:
Sure.